Source/Core/VideoCommon/ShaderCache.cpp

```diff
@@ -0,0 +1,909 @@
+// Copyright 2018 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "VideoCommon/ShaderCache.h"
+
+#include "Common/Assert.h"
+#include "Common/MsgHandler.h"
+#include "Core/Host.h"
+
+#include "VideoCommon/RenderBase.h"
+#include "VideoCommon/Statistics.h"
+#include "VideoCommon/VertexLoaderManager.h"
+#include "VideoCommon/VertexManagerBase.h"
+
+std::unique_ptr<VideoCommon::ShaderCache> g_shader_cache;
+
+namespace VideoCommon
+{
+ShaderCache::ShaderCache() = default;
+ShaderCache::~ShaderCache() = default;
+
+bool ShaderCache::Initialize()
+{
+  m_api_type = g_ActiveConfig.backend_info.api_type;
+  m_host_config = ShaderHostConfig::GetCurrent();
+  m_efb_multisamples = g_ActiveConfig.iMultisamples;
+
+  // Create the async compiler, and start the worker threads.
+  m_async_shader_compiler = std::make_unique<VideoCommon::AsyncShaderCompiler>();
+  m_async_shader_compiler->ResizeWorkerThreads(g_ActiveConfig.GetShaderPrecompilerThreads());
+
+  // Load shader and UID caches.
+  if (g_ActiveConfig.bShaderCache)
+  {
+    LoadShaderCaches();
+    LoadPipelineUIDCache();
+  }
+
+  // Queue ubershader precompiling if required.
+  if (g_ActiveConfig.CanPrecompileUberShaders())
+    PrecompileUberShaders();
+
+  // Compile all known UIDs.
+  CompileMissingPipelines();
+
+  // Switch to the runtime shader compiler thread configuration.
+  m_async_shader_compiler->ResizeWorkerThreads(g_ActiveConfig.GetShaderCompilerThreads());
+  return true;
+}
+
+void ShaderCache::SetHostConfig(const ShaderHostConfig& host_config, u32 efb_multisamples)
+{
+  if (m_host_config.bits == host_config.bits && m_efb_multisamples == efb_multisamples)
+    return;
+
+  m_host_config = host_config;
+  m_efb_multisamples = efb_multisamples;
+  Reload();
+}
+
+void ShaderCache::Reload()
+{
+  m_async_shader_compiler->WaitUntilCompletion();
+  m_async_shader_compiler->RetrieveWorkItems();
+
+  InvalidateCachedPipelines();
+  ClearShaderCaches();
+
+  if (g_ActiveConfig.bShaderCache)
+    LoadShaderCaches();
+
+  // Switch to the precompiling shader configuration while we rebuild.
+  m_async_shader_compiler->ResizeWorkerThreads(g_ActiveConfig.GetShaderPrecompilerThreads());
+  CompileMissingPipelines();
```

Comment:
Shouldn't this also recompile the ubershaders? If so, worth refactoring with Init()?

Reply:
As the ubershaders UIDs are already in the map (done at Init time), the shaders will be indirectly recompiled when the corresponding pipelines are created. It was unclear, so I've added a comment to clarify this does happen.